### ISSUE TITLE
Explain discriminator loss?

### ISSUE THREAD

**Issue by ovcharenkoo** [shepnerd/inpainting_gmcnn#51]
Hi @shepnerd, thanks for the interesting work! 

I have a hard time understanding the logic of discriminator. Could you please explain the motivation for the value range of discriminator outputs and respective loss?

Here is my understanding:
The step of evaluation of 'realism' for inpainted and ground truth images is ([here](https://github.com/shepnerd/inpainting_gmcnn/blob/ba7f7109c38c3805800283cdb9d79cd7c4a3294f/pytorch/model/net.py#L304), excluding the local terms for brevity):

`self.completed_logit, _ = self.netD(self.completed.detach(), self.completed_local.detach())`
`self.gt_logit, _ = self.netD(self.gt, self.gt_local)`

where  `self.completed_logit`  and `self.gt_logit` are outputs from the last linear layer of the discriminator and these contain a single value for each image in the batch. 

So the discriminator does mapping `[b, c, h, w] --> [b, 1]`. This part is clear, except for the range of output values which might be arbitrary due to the [linear activation in the last layer of discriminator](https://github.com/shepnerd/inpainting_gmcnn/blob/ba7f7109c38c3805800283cdb9d79cd7c4a3294f/pytorch/model/net.py#L184). So it might be -3, 5, 0.3, etc.

The loss term is defined [as follows](https://github.com/shepnerd/inpainting_gmcnn/blob/ba7f7109c38c3805800283cdb9d79cd7c4a3294f/pytorch/model/net.py#L308):
`self.D_loss = nn.ReLU()(1.0 - self.gt_logit).mean() + nn.ReLU()(1.0 + self.completed_logit).mean()`

As I understand, a perfect discriminator should assign  anything **above 1 to real image**, `self.gt_logit` (identifies the ground truth as real) and anything **less than -1 to fake image**,  `self.completed_logit` (identifies the inpainted image as fake).

Is that correct? If so, could you explain what is the practical difference in using these compared to regular 0 and 1 adversarial ground truths like [here](https://github.com/eriklindernoren/PyTorch-GAN/blob/a163b82beff3d01688d8315a3fd39080400e7c01/implementations/pix2pix/pix2pix.py#L131)?

Thanks!

**Comment by shepnerd** [comment]
Your understanding for the mentioned code is correct. How to measure the realness and fakeness of the generated images or real ones is a broad research field. The well-known measurements used in it includes Squared Euclidean distance (LSGAN, your given link), cross entropy (the original one used in the first GAN paper), Wasserstein-1 metric (WGAN and its variants), and so on. For the used labels in LSGAN, you can refer to this [blog](https://medium.com/@jonathan_hui/gan-lsgan-how-to-be-a-good-helper-62ff52dd3578) or the original [paper](https://arxiv.org/pdf/1611.04076.pdf) for understanding.

**Comment by ovcharenkoo** [comment]
Thank you for your prompt reply! The LSGAN is something new for me
